### PR TITLE
Type annotations for the rest of the core modules

### PR DESCRIPTION
**Pull request by magnatelee**
This PR adds type annotations to the following modules:

* `legate.core.runtime`
* `legate.core.store`
* `legate.core.solver`

### CODE REVIEW

File: legate/core/launcher.py [hunk 1]
```diff
@@ -744,13 +750,19 @@ def add_store(
...
+
             has_storage = perm != Permission.WRITE
             read_only = perm == Permission.READ
             if has_storage:
                 self.add_future(store.storage)
             args.append(FutureStoreArg(store, read_only, has_storage, redop))
 
         else:
+            if TYPE_CHECKING:
+                assert isinstance(store.storage, RegionField)
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/launcher.py:764]: `RegionField` is imported only when `TYPE_CHECKING` is true, so checking this assertion unconditionally doesn't work. there are several places like this I believe. I guess the right fix is to hoist `RegionField` and other primitives to their own modules so we can import them unconditionally without cyclic dependencies.

File: legate/core/partition.py [hunk 2]
```diff
@@ -40,27 +41,61 @@ class Restriction(IntEnum):
...
+        self, restrictions: Sequence[Restriction]
+    ) -> bool:
+        ...
+
+    @abstractmethod
+    def needs_delinearization(self, launch_ndim: int) -> bool:
+        ...
+
+    @abstractproperty
+    def requirement(self) -> Union[Type[Broadcast], Type[Partition]]:
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/partition.py:78]: sounds like a good idea, though I'll probably name it `RequirementType`

File: legate/core/_legion/task.py [hunk 3]
```diff
@@ -1017,7 +1020,11 @@ def set_sharding_space(self, space: IndexSpace) -> None:
 
     @dispatch
     def launch(
-        self, runtime: Runtime, context: Context, redop: int = 0
+        self,
+        runtime: legion.legion_runtime_t,
+        context: legion.legion_context_t,
+        redop: int = 0,
+        **kwargs: Any,
     ) -> Union[Future, FutureMap]:
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/_legion/task.py:1028]: e95fa8446cfe4f7b3993c7e0fdb0a6bfa2c56026

File: legate/core/utils.py [hunk 4]
```diff
@@ -62,6 +62,13 @@ def __contains__(self, obj: object) -> bool:
     def __iter__(self) -> Iterator[T]:
         return iter(self._dict)
 
+    def remove_all(self, other: OrderedSet[T]) -> OrderedSet[T]:
+        result: OrderedSet[T] = OrderedSet()
+        for obj in self:
+            if obj not in other:
+                result.add(obj)
+        return result
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/utils.py]: 622654ab5ebf894f4157bcb0ddb881b165bedf1a

File: legate/core/partition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/partition.py]: cff5203fff76f22b1a590498ed725feca04f28f3

**Comment by bryevdv** [comment]
LGTM!

**Review by bryevdv — APPROVED** [review]

**Pull request merged**